tests/test_grad.py

```diff
@@ -0,0 +1,415 @@
+def test_prim_grads(prim, cases):
+    for case in cases:
+        _grad_test(pyi[prim], prim, case)
+
+
+def grad_test(*tests):
+    """Decorate a function to parse and run it against pure Python.
+
+    Returns a unit test that will parse the function, and then for
+    each `inputs` tuple in `tests` it will check that the pure Python,
+    undecorated function returns that same output.
+
+    Arguments:
+        tests: One or more inputs tuple.
+
+    """
+
+    def decorate(fn):
+        def test(args):
+            if not isinstance(args, tuple):
+                args = (args,)
+
+            _grad_test(fn, fn, args)
+
+        m = pytest.mark.parametrize('args', list(tests))(test)
+        m.__orig__ = fn
+        return m
+    return decorate
+
+
+@grad_test((13.0, 14.0))
+def test_null(x, y):
+    """Test null gradient."""
+    return 10.0 + 28.0 / 43.0
+
+
+@grad_test((1.0, 4.0), (5.0, -13.0))
+def test_grad_add(x, y):
+    return x + y
+
+
+@grad_test((3.0, 4.0))
+def test_grad_expr(x, y):
+    return x**3.0 * y**4.0
+
+
+@grad_test((3.0,))
+def test_constant(x):
+    """Test the use of a literal in the expression."""
+    return 18.0 * x
+
+
+@grad_test((3.0,))
+def test_dup_args_in_call(x):
+    """The naive gradient update rule fails when a function's arguments
+    contain the same variable more than once."""
+    return x * x
+
+
+@grad_test((3.0,))
+def test_quadruple_args_in_call(x):
+    """Test that duplicated arguments still cause no problem even if
+    there are four of them."""
+    def g(a, b, c, d):
+        return a * b * c * d
+    return g(x, x, x, x)
+
+
+@grad_test((3.0, 5.0))
+def test_tuples(x, y):
+    tup = x + y, x * y
+    z = tup[0] + tup[1]
+    return z
+
+
+@grad_test((4.0, 5.0))
+def test_hof(a, b):
+    """Test higher order functions."""
+    def f(g, x):
+        return g(x) * g(x + 10.0)
+
+    def g(x):
+        return x * b
+
+    return f(g, a) + f(g, b)
+
+
+@grad_test((4.0, 5.0))
+def test_hof_tup(a, b):
+    """Test higher order functions."""
+    def f(gh, x, y):
+        g, h = gh
+        return g(x, y) * h(x, y)
+
+    return f((scalar_add, scalar_mul), a, b)
+
+
+@grad_test((4.0, 5.0))
+def test_simple_closure(a, b):
+    """Test some trivial closures."""
+    def f():
+        return a + 1.0
+
+    def g():
+        return b + 2.0
+    return f() * g()
+
+
+@grad_test((4.0,))
+def test_closure(a):
+    """This is the closure test in the paper."""
+    def x1(b):
+
+        def x4(c):
+            return b
+        return x4
+    x2 = x1(a)
+    x3 = x2(1.0)
+    return x3
+
+
+@grad_test((4.0, 5.0), (6.4, -7.8))
+def test_if(a, b):
+    # This is max, but what this is really testing is the most basic
+    # if statement, so I prefer to name the test 'test_if'
+    if a > b:
+        return a
+    else:
+        return b
+
+
+@grad_test((4.0, 5.0), (6.4, -7.8))
+def test_if2(a, b):
+    # This is max, but what this is really testing is the most basic
```

Comment:
This is not really `max` any more, I think :)

Reply:
Indeed, it is not.